OIDC RequireHttpsMetadata

Tratcher opened the pull request:
#485 
Something like this @PinpointTownes?
@muratg @Eilon @brentschmaltz @tushargupta51 @blowdart 

(I'll do Bearer once we agree on the fix).

dnfclas commented:
Hi __@Tratcher__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

PinpointTownes commented:
:shipit: it looks fine.

3 remarks/questions though:
* I think it would be appropriate to update the OIDC middleware to throw an exception when a non-HTTPS authority/metadata endpoint is used without `RequireHttpsMetadata` being set to `false` directly at the startup time (the document retriever will only throw an exception during the challenge, it's too late IMHO).

* Should we automatically set it to `false` on production/testing environments, similarly to what you've done in Hosting to make the development stage smoother?

* Should we generalize this property to related things like the token request or the callback stage to make sure the OIDC middleware doesn't send insecure requests and doesn't listen on non-HTTPS endpoints?

Eilon commented on src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectOptions.cs:
```diff
@@ -106,6 +106,12 @@ public string DisplayName
         public string ClientSecret { get; set; }
 
         /// <summary>
+        /// Gets or sets if https is required for the metadata address or authority.
```
Make `HTTPS` uppercase in the comment.

Eilon commented on src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectOptions.cs:
```diff
@@ -106,6 +106,12 @@ public string DisplayName
 
         /// <summary>
+        /// Gets or sets if https is required for the metadata address or authority.
+        /// The default is true. This should only be disabled in development environments.
```
Grammar police: `This should be disabled only in development environments.` :monkey_face:

PinpointTownes commented:
Is this still an option we want to add?

Tratcher commented:
Yes, I'll get back to it this week.

Tratcher commented:
Rebased and updated.
@PinpointTownes I think there's a big difference between requiring metadata retrieval to use https vs requiring your site to run on https. I have almost never needed to get my metadata from an http source, but I always develop my sites using http because https is harder to set up locally.

PinpointTownes commented:
> @PinpointTownes I think there's a big difference between requiring metadata retrieval to use https vs requiring your site to run on https.

Not from a security perspective. @brentschmaltz claimed that this change was to make the OIDC stuff "secure by default". If only the initial stage requires HTTPS, then it's not really "secure by default", as the whole authentication flow (authorization response, token request/response) would need to use HTTPS to avoid disclosing critical data like client credentials or access/identity/refresh tokens.

> I have almost never needed to get my metadata from an http source, but I always develop my sites using http because https is harder to set up locally.

I guess that's because you're not using your own authorization/identity server, but I bet that most IdentityServer3/ASOS users will develop their authorization server using HTTP.

PinpointTownes commented:
Thanks for adding the early check :+1:

HaoK commented:
:shipit:
